File: src/commands/rtp.ts
import { createCommand } from ':util/creators.ts';
import { getSubcommand } from ':util/commands.ts';
import { replyToInteraction } from ':util/interactions.ts';
import {
    getRTPMembers,
    addMemberToRTP,
    removeMemberFromRTP,
} from ':util/rtp.ts';
import { log } from ':util/logger.ts';
import { SubcommandMap } from ':interfaces/command.ts';
import {
    ApplicationCommandTypes,
    ApplicationCommandOptionTypes,
    Interaction,
} from 'discordeno';

createCommand({
    name: 'rtp',
    description: '🥍 Manage RTP status',
    type: ApplicationCommandTypes.ChatInput,

    options: [
        {
            name: 'join',
            description: '🟢 Join RTP',
            type: ApplicationCommandOptionTypes.SubCommand,
            required: false,
        },
        {
            name: 'leave',
            description: '⭕ Leave RTP',
            type: ApplicationCommandOptionTypes.SubCommand,
            required: false,
        },
        {
            name: 'info',
            description: "👀 Check out who's available",
            type: ApplicationCommandOptionTypes.SubCommand,
            required: false,
        },
    ],

    run: async (interaction) => {
        const subcommand = getSubcommand(interaction);

        if (!subcommand) {
            throw new Error('Cannot execute /rtp without a subcommand');
        }

        const subcommands: SubcommandMap = {
            join: rtpJoin,
            leave: rtpLeave,
            info: rtpInfo,
        };

        await subcommands[subcommand](interaction);
    },
});

/** Function for `/rtp join`. */
const rtpJoin = async (interaction: Interaction) => {
    const member = interaction.member!;
    await addMemberToRTP(member);

    const name = member.nick || interaction.user.username;

    await replyToInteraction(interaction, {
        content: `🟢  ${name} is now available for Lunaro`,
    });
};

/** Function for `/rtp leave`. */
const rtpLeave = async (interaction: Interaction) => {
    const member = interaction.member!;
    await removeMemberFromRTP(member);

    const name = member.nick || interaction.user.username;

    await replyToInteraction(interaction, {
        content: `⭕  ${name} is no longer available for Lunaro`,
    });
};

/** Function for `/rtp info`. */
const rtpInfo = async (interaction: Interaction) => {
    const rtpMembers = await getRTPMembers();
    const rtpMemberCount = rtpMembers.length;

    const rtpMemberCountString =
        rtpMemberCount === 1
            ? 'There is 1 member'
            : `There are ${rtpMemberCount} members`;

    await replyToInteraction(interaction, {
        content: `👀  ${rtpMemberCountString} available for Lunaro`,
    });

    log(`${rtpMemberCountString} with the RTP role`);
};

File: src/commands/rtp.rs
use poise::{command, CreateReply};

use crate::{
    env::Environment,
    types::{error::Error, poise::PoiseContext},
    util::rtp,
};

/// 🥍 Manage your RTP status
#[command(slash_command, rename = "rtp", subcommands("join", "leave", "check"))]
pub async fn run(_context: PoiseContext<'_>) -> Result<(), Error> {
    // Handled in subcommands
    Ok(())
}

/// 🟢 Equip your Arcata
#[command(slash_command)]
async fn join(context: PoiseContext<'_>) -> Result<(), Error> {
    let env = Environment::instance();

    let member = &mut context
        .serenity_context()
        .http
        .get_member(env.home_guild_id.into(), context.author().id)
        .await?;

    rtp::add(member, context.serenity_context()).await?;

    let display_name = match &member.nick {
        Some(nick) => nick,
        None => &member.user.name,
    };

    context
        .send(
            CreateReply::default()
                .content(format!("🟢  {display_name} is now available for Lunaro")),
        )
        .await?;

    Ok(())
}

/// ⭕ Unequip your Arcata
#[command(slash_command)]
async fn leave(context: PoiseContext<'_>) -> Result<(), Error> {
    let env = Environment::instance();

    let member = &mut context
        .serenity_context()
        .http
        .get_member(env.home_guild_id, context.author().id)
        .await?;

    let display_name = match &member.nick {
        Some(nick) => nick,
        None => &member.user.name,
    };

    context
        .send(CreateReply::default().content(format!(
            "⭕  {display_name} is no longer available for Lunaro"
        )))
        .await?;

    rtp::remove(member, context.serenity_context()).await?;

    Ok(())
}

/// 👀 Check who's equipped their Arcata
#[command(slash_command)]
async fn check(context: PoiseContext<'_>) -> Result<(), Error> {
    let ready_member_count = rtp::count(context.serenity_context()).await?;

    let verb = match ready_member_count {
        1 => "is",
        _ => "are",
    };

    context
        .send(CreateReply::default().content(format!(
            "👀  There {verb} {ready_member_count} Tenno available for Lunaro"
        )))
        .await?;

    Ok(())
}
